### PR TITLE
Handle TypeVar types and enables most common generics support

### LINKED ISSUE #91: Support typing.TypeVar annotations
Hamilton fails on TypeVar annotations.

## Current behavior
If you define functions like the following:
```python
import typing

StrSeries = typing.TypeVar("StrSeries", bound="pd.core.series.Series[str]")

def normalize_situs(situs_addr: StrSeries) -> StrSeries:
    normalized = situs_addr.map(str.upper)
    return normalized

def foo_bar(normalize_situs: StrSeries) -> StrSeries:
    return normalize_situs
```
and then try to create a DAG it breaks.

## Stack Traces
```python
Traceback (most recent call last):
  File "sf/hamilton/examples/temp/run.py", line 21, in <module>
    dr = driver.Driver({'situs_addr': pd.Series(['a', 'b', 'c'])}, repo)
  File "sf/hamilton/hamilton/driver.py", line 47, in __init__
    self.graph = graph.FunctionGraph(*modules, config=config, adapter=adapter)
  File "sf/hamilton/hamilton/graph.py", line 183, in __init__
    self.nodes = create_function_graph(*modules, config=self._config, adapter=adapter)
  File "sf/hamilton/hamilton/graph.py", line 117, in create_function_graph
    add_dependency(n, node_name, nodes, param_name, param_type, adapter)
  File "sf/hamilton/hamilton/graph.py", line 78, in add_dependency
    if custom_subclass_check(required_node.type, param_type):
  File "sf/hamilton/hamilton/graph.py", line 41, in custom_subclass_check
    if issubclass(requested_type, param_type):
TypeError: issubclass() arg 1 must be a class
```

## Expected behavior
Hamilton should be happy to handle TypeVar defined types.

## Library & System Information
E.g. python version, hamilton library version, linux, etc.

**Steps to replicate behavior**
1.  define functions above into module.
2.  Run
```
import repo
dr = driver.Driver({'situs_addr': pd.Series(['a', 'b', 'c'])}, repo)
```

## Additional context
Flagged by discord user https://discord.com/channels/842466558483496997/842467113611821087/955688991749980200

### PR DESCRIPTION
**Pull request by skrawcz**
High level:

- Fixes #91 (handling TypeVar types)
- Implements #39  (adding support for generic typing)

## Changes

- Refactors code to do type checks. This could look nicer, but it gets the job done. Unit tests don't break.
- With the refactor I added support for generics. Tests added for this. There are edge cases I did not handle. E.g. didn't check TypedDicts.
- 

## Testing

1. Added unit tests.
2. Will see what CI tests break and see if I need to iterate.

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [x] Code follows the standards laid out in the 
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Python

- [x] python 3.6
- [x] python 3.7

### CODE REVIEW

File: tests/test_graph.py [hunk 1]
```diff
@@ -417,3 +419,83 @@ def test_combine_inputs_collision_2():
...
+
+class Y(X):
+    pass
+
+
+custom_type = typing.TypeVar('FOOBAR')
+
+
+@pytest.mark.parametrize('param_type,required_type,expected', [
+    (custom_type, custom_type, True),
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply tests/test_graph.py:436]: yep was thinking I need to add a few more cases.

[the remark this answers is not in the document]
Comment (skrawcz) [reply tests/test_graph.py:436]: actually I think I have them mostly covered. Not sure what else to add.

**Review by elijahbenizzy — APPROVED** [review]

File: tests/test_graph.py [hunk 2]
```diff
@@ -417,3 +419,83 @@ def test_combine_inputs_collision_2():
...
+    (typing.List[int], typing.List, True),
+    (typing.List, typing.List[float], True),
+    (typing.List, list, True),
+    (typing.Dict, dict, True),
+    (dict, typing.Dict, True),
+    (list, typing.List, True),
+    (list, typing.List, True),
+    (typing.List[int], typing.List[float], False),
+    (typing.Dict, typing.List, False),
+    (typing.Mapping, typing.Dict, True),
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply tests/test_graph.py:449]: the reverse should be `False` right?

File: tests/test_graph.py [hunk 3]
```diff
@@ -417,3 +419,83 @@ def test_combine_inputs_collision_2():
...
+    (typing.Iterable, typing.List, True),
+    (typing.Tuple[str, str], typing.Tuple[str, str], True),
+    (typing.Tuple[str, str], typing.Tuple[str], False),
+    (typing.Tuple[str, str], typing.Tuple, True),
+    (typing.Tuple, typing.Tuple[str, str], True),
+    (typing.Union[str, str], typing.Union[str, str], True),
+    (X, X, True),
+    (X, Y, True),
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply tests/test_graph.py:457]: sure

File: tests/test_graph.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (skrawcz) [reply tests/test_graph.py:449]: added both.
